cognite_toolkit/_cdf_tk/utils/fileio/_readers.py

```diff
@@ -75,34 +79,130 @@ class YMLReader(YAMLBaseReader):
-            yield {key: self._parse_value(value) for key, value in row.items()}
+    def __init__(
+        self,
+        input_file: Path,
+        sniff_rows: int | None = None,
+        schema: Sequence[SchemaColumn] | None = None,
+        keep_failed_cells: bool = False,
+    ) -> None:
+        super().__init__(input_file)
+        if sniff_rows is not None and schema is not None:
+            raise ValueError("Cannot specify both `sniff_rows` and `schema`. Use one or the other.")
+        elif sniff_rows is not None and schema is None:
+            self.schema: Sequence[SchemaColumn] | None = self.sniff_schema(input_file, sniff_rows)
+        else:
+            self.schema = schema
+        self.sniff_rows = sniff_rows
+        self.keep_failed_cells = keep_failed_cells
+        self.parse_function_by_column = self._create_parse_functions(self.schema)
+        self.failed_cell: list[FailedParsing] = []
+
+    @classmethod
+    def _create_parse_functions(
+        cls, schema: Sequence[SchemaColumn] | None
+    ) -> dict[str, Callable[[str | None], JsonVal]]:
+        """Create a dictionary of parse functions for each column in the schema."""
+        parse_function_by_column: dict[str, Callable[[str | None], JsonVal]] = defaultdict(
+            lambda: cls._default_parse_function
+        )
+        if schema is not None:
+            for column in schema:
+                if column.type in {"date", "timestamp"}:
+                    raise ToolkitValueError("CSVReader does not support 'date' or 'timestamp' types.")
+                parse_function_by_column[column.name] = partial(  # type: ignore[assignment]
+                    convert_str_to_data_type, type_=column.type, nullable=True, is_array=False
+                )
+        return parse_function_by_column
 
     @staticmethod
-    def _parse_value(value: str) -> JsonVal:
-        """Parse a string value into its appropriate type."""
-        if value == "":
+    def _default_parse_function(value: str | None) -> JsonVal:
+        if value is None:
             return None
-        try:
-            # Try parsing as JSON (for lists, dicts, true, false, null, numbers)
-            return json.loads(value)
-        except (json.JSONDecodeError, TypeError):
-            pass
-        try:
-            return int(value)
-        except ValueError:
-            pass
-        try:
-            return float(value)
-        except ValueError:
-            pass
-        if value.lower() in ("true", "false"):
-            return value.lower() == "true"
-        return value
+        return infer_data_type_from_value(value, "Json")[1]
+
+    @classmethod
+    def sniff_schema(cls, input_file: Path, sniff_rows: int) -> list[SchemaColumn]:
+        """Sniff the schema from the first `sniff_rows` rows of the CSV file."""
+        if sniff_rows <= 0:
+            raise ValueError("`sniff_rows` must be a positive integer.")
+
+        with input_file.open("r", encoding="utf-8") as file:
```

Comment:
Maybe I got this whole thing wrong, but are you doing file open operation twice here? I mean, when you are going instantiate CSVReader, you are opening the file, and then reopening it again here to "sniff the schema"?